message_ix/macro.py

```diff
@@ -1,61 +1,399 @@
+    # production function
+    # exponent rho (esub is elasticity of substitution  between energy and the
+    # value added aggregates, i.e. capital and labor)
+    rho = (dict['esub'] - 1) / dict['esub']
+    # production Y in base year
+    gdp0 = gdp_calibrate.loc[dict['baseyearmacro']]
+    # capital K in base year
+    k0 = dict['kgdp'] * gdp0
+    # investments I in base year
+    i0 = k0 * (dict['growth'].loc[dict['firstmodelyear']] + dict['depr'])
+    # consumption C in base year
+    c0 = gdp0 - i0 - total_cost.loc[dict['baseyearmacro']]
+
+    # production function coefficients for capital, labor (aconst) and
+    # energy (bconst) are calibrated from the first-order optimality
+    # condition, i.e. bconst from dY/d"ELEC" = peref or dY/d"NELE" = pnref
+    # (equivalent as shown numerically below) and aconst by inserting bconst
+    # back into the production function, setting the labor force index in the
+    # base year to 1 (numeraire) and solving for aconst
+
+    # production function coefficients of the different energy sectors
+    # (factor of 1000 needed for billion US. to trillion US. conversion)
+    dict['demand_ref'] = dict['demand_ref'].pivot_table(columns='commodity',
+                                                        index='node',
+                                                        values='lvl')
+    dict['p_ref'] = dict['p_ref'].pivot_table(columns='commodity',
+                                              index='node',
+                                              values='lvl')
+
+    def calc_bconst(row):
+        val = (dict['p_ref'].loc[row.name] / 1e3 * (
+                float(1 / gdp0.loc[row.name]) * row) ** float(
+            rho.loc[row.name] - 1))
+        return val
+
+    bconst = dict['demand_ref'].apply(calc_bconst, axis=1)
+
+    # production function coefficient of capital and labor
+    # (factor of 1000 needed for billion US. to trillion US. conversion)
+    def calc_partmp(row):
+        val = bconst.loc[row.name] * row ** rho.loc[row.name]
+        return val
+
+    partmp = dict['demand_ref'].apply(calc_partmp, axis=1)
+
+    def calc_aconst(row):
+        val = ((gdp_calibrate.loc[dict['baseyearmacro'], row.name] / 1000) **
+               rho.loc[
+                   row.name] - partmp.sum(axis=0) / (
+                           (k0.loc[row.name] / 1000) ** (
+                           rho.loc[row.name] * dict['kgdp'].loc[row.name])))
+        return val
+
+    aconst = partmp.apply(calc_aconst, axis=1)
+
+    # ------------------------------------------------------------------------
+    # add sets needed for MACRO (if they don't exist)
+    # ------------------------------------------------------------------------
+    # sets that are standard part of a MESSAGE scenario, but needed for
+    # running MACRO standalone
+    if not new.has_set("node"):
+        new.init_set("node")
+    for node in dict['region_list']:
```

Comment:
I am not sure how this "running MACRO standalone" thing is supposed to work so I am not sure what the impact of removing this would be. Should I still simply remove it? 

I now introduced a kwarg `macro_regions` and changed the code to:
```suggestion
    for node in macro_regions:
        new.add_set("node", node)
```

Reply:
@behnam2015 says he never needed that to get started, so let’s remove this feature

Reply:
@danielhuppmann, you suggest we delete all regular "MESSAGE sets" (i.e.: `node`, `cat_node`, `year`, `cat_year` and `commodity` ) from the `MACRO_SPECS['sets']`?